content/blog/0007-signals.md

```diff
@@ -0,0 +1,362 @@
+* Signals have a clock domain associated with them
+
+Hence, if we look at the definition of Clash's `Signal` we see:
+
+```haskell
+data Signal (dom :: Domain) (a :: Type) = a :- Signal dom a
+```
+
+We can use this definition in simulation:
+
+```haskell
+>>> signal = pure 1 :: Signal System Int
+>>> signal
+1 :- 1 :- 1 :- 1 :- 1 :-
+^C
+```
+
+To save some time, I pressed interrupt (`^C`) to prevent Haskell from printing indefinitely. Like lists, we can map over it:
+
+```haskell
+>>> fmap (+10) signal -- More typically written as `(+10) <$> signal`
+11 :- 11 :- 11 :- 11 :- 11 :-
+^C
+```
+
+We can also combine multiple signals and perform some computation on them<sup><a href="#footnote-2">2</a></sup>:
+
+```haskell
+>>> liftA2 (+) signal signal -- More typically written as `(+) <$> signal <*> signal`
+2 :- 2 :- 2 :- 2 :- 2 :-
+^C
+```
+
+As discussed in the _combinational logic_ section, Clash shouldn't be able to translate the definitions of `fmap` and `liftA2`: after all, they act on infinitely recurring data types! In practice it can, because both `pure` and these two operations are marked as _primitive_: Clash will not try to translate the definitions of primitives, it will generate hardcoded bits of HDL.
+
+### Signals 🤝 state
+So far we've only seen combinational operations: we haven't introduced any state yet. Worse, with the introduced primitives we cannot add any state either. Let's take a step back and see what it would have to look like:
+
+<script type="wavedrom">
+{ signal: [
+  { name: "clock", wave: '0p...' },
+  { name: "D",    wave: "22222",  data: ["A", "B", "C", "D", "E"] },
+  { name: "Q",    wave: "22222",  data: ["init", "A", "B", "C", "D", "E"] },
+]}
+</script>
+
+I.e., the output signal **Q** is a shifted version of **D**. In this diagram, the initial state of the register is marked **init**. A fourth primitive Clash offers is therefore `register`:
+
+```haskell
+-- | Simplified version of Clash's `register`: the actual version has more control signals.
+register :: Clock dom -> a -> Signal dom a -> Signal dom a
+register _clk init signal = init :- signal
+```
+
+Again, when Clash sees this primitive used, it generates a hardcoded bit of HDL. Note that because `register` operates on a single domain, it does not need to be aware of any properties of the clock; the abstraction works the same for clocks running on 1 MHz or 100 MHz. The generated HDL *does* need clock lines to be routed though, so the primitive still needs the clock argument.
+
+And just like that, we've introduced state. We can now implement a simple counter:
+
+```haskell
+counter = register clock (fmap (+1) counter)
+
+-- Because `Signal` has a `Num` instance, we can also write:
+-- counter = register clock (counter + 1)
+```
+
+From an HDL compilation point of view, we can see how this might work. If Clash encounters the special primitive `register`, it will not try to translate the definition, but it will insert a predefined bit of HDL.
```

Comment:
```suggestion
From an HDL compilation point of view, we can see how this might work. If Clash encounters the special primitive [`register`](https://hackage.haskell.org/package/clash-prelude-1.8.2/docs/Clash-Explicit-Signal.html#v:register), it will not try to translate the definition, but it will insert a predefined bit of HDL that implements the register.
```